### PR TITLE
Added new 6 requests with more stuff

### PR DESCRIPTION
**Pull request by acceler8tion**
# Preview

- Added new **6** requests & methods ( *Authenticated* : **6**, *Anonymous* : **0** )
- Added **SHA1** encryption, more stuff in [Utils](https://github.com/acceler8tion/jdash/blob/master/src/main/java/com/github/alex1304/jdash/util/Utils.java)
- Added new **URLs**

---

# New GDRequest

- Added new 6 requests

|Name|Method|RequestType|Javadoc|Test|Issue Code|
|:---|:---|:---:|:---:|:---:|:---:|
|**GDBlockedUsersRequest** [🔗](https://github.com/acceler8tion/jdash/blob/master/src/main/java/com/github/alex1304/jdash/client/GDBlockedUsersRequest.java)|[getBlockedUsers](https://github.com/acceler8tion/jdash/blob/master/src/main/java/com/github/alex1304/jdash/client/AuthenticatedGDClient.java#L172)|**Authenticated**|:white_check_mark:|Successful|:x:|
|**GDFriendListRequest** [🔗](https://github.com/acceler8tion/jdash/blob/master/src/main/java/com/github/alex1304/jdash/client/GDFriendListRequest.java)|[getPrivateFriends](https://github.com/acceler8tion/jdash/blob/master/src/main/java/com/github/alex1304/jdash/client/AuthenticatedGDClient.java#L211)|**Authenticated**|:white_check_mark:|Successful|:x:|
|**GDLevelDemonRatingRequest** [🔗](https://github.com/acceler8tion/jdash/blob/master/src/main/java/com/github/alex1304/jdash/client/GDLevelDemonRatingRequest.java)|[rateDemon](https://github.com/acceler8tion/jdash/blob/master/src/main/java/com/github/alex1304/jdash/client/AuthenticatedGDClient.java#L142)|**Authenticated**|:white_check_mark:|Failed|**1, 2**|
|**GDLevelStarsRatingRequest** [🔗](https://github.com/acceler8tion/jdash/blob/master/src/main/java/com/github/alex1304/jdash/client/GDLevelStarsRatingRequest.java)|[rateStars](https://github.com/acceler8tion/jdash/blob/master/src/main/java/com/github/alex1304/jdash/client/AuthenticatedGDClient.java#L124)|**Authenticated**|:white_check_mark:|**90%** Success|**1**|
|**GDUserBlockRequest** [🔗](https://github.com/acceler8tion/jdash/blob/master/src/main/java/com/github/alex1304/jdash/client/GDUserBlockRequest.java)|[blockUser](https://github.com/acceler8tion/jdash/blob/master/src/main/java/com/github/alex1304/jdash/client/AuthenticatedGDClient.java#L183)|**Authenticated**|:white_check_mark:|Successful|:x:|
|**GDUserUnblockRequest** [🔗](https://github.com/acceler8tion/jdash/blob/master/src/main/java/com/github/alex1304/jdash/client/GDUserUnblockRequest.java)|[unblockUser](https://github.com/acceler8tion/jdash/blob/master/src/main/java/com/github/alex1304/jdash/client/AuthenticatedGDClient.java#L197)|**Authenticated**|:white_check_mark:|Successful|:x:|

> **Note!** The issues are at the bottom.

---

# Utils

- Added sha1, some stuff

|Method|Desc|
|:---|:---|
|**sha1Encrypt** [🔗](https://github.com/acceler8tion/jdash/blob/master/src/main/java/com/github/alex1304/jdash/util/Utils.java#L142)|Do sha1 encryption.(*I think it needs renaming.*)|
|**randomString** [🔗](https://github.com/acceler8tion/jdash/blob/master/src/main/java/com/github/alex1304/jdash/util/Utils.java#L157)|Generates a random string as long as the size.|
|**buildChk** [🔗](https://github.com/acceler8tion/jdash/blob/master/src/main/java/com/github/alex1304/jdash/util/Utils.java#L166)|Build a chk(*idk what chk means..*)|

> **Note!** The issues are at the bottom.

---

# Issues

1. There is no operational problem, but I am not sure if it applies to the actual server.
> Despite the incorrect param input during the test, I've seen a 1 in response, and it's a personal experience, but I've also made a "level like" request before, and there was a similar phenomenon (which was used as sha1, xor) and it didn't apply to the actual server.

2. There is no problem with body value configuration (checked with packet capture), but due to "500 Internal error", i can't get successful result from test.

3. As we talked last time, i didn't add a level reporting function yet.

4. I've never made a friend acceptance, rejection, and pending list of friends, so unfortunately, I couldn't include it.

# Contact

Discord: Acceler8tionAsDF#**3350**

### CODE REVIEW

File: src/main/java/com/github/alex1304/jdash/client/AbstractGDClient.java
Comment (Alex1304) [inline src/main/java/com/github/alex1304/jdash/client/AbstractGDClient.java:58]: Do not remove this line, it was intentionally added to prevent the server from blocking the requests (see 3.3.8 release patch notes)

File: src/main/java/com/github/alex1304/jdash/client/AuthenticatedGDClient.java
Comment (Alex1304) [inline src/main/java/com/github/alex1304/jdash/client/AuthenticatedGDClient.java:218]: * "private friends" is a weird name, why not just `getFriends()`?
* If the server returns info on the pagination of the friends list, please return a `Mono<GDPaginator>` instead of a `Mono<List>`.

File: src/main/java/com/github/alex1304/jdash/client/GDLevelDemonRatingRequest.java
Comment (Alex1304) [inline src/main/java/com/github/alex1304/jdash/client/GDLevelDemonRatingRequest.java:36]: Create a new subclass of `GDClientException` in the `com.github.alex1304.jdash.exception` package.

File: src/main/java/com/github/alex1304/jdash/client/GDLevelStarsRatingRequest.java
Comment (Alex1304) [inline src/main/java/com/github/alex1304/jdash/client/GDLevelStarsRatingRequest.java:46]: Create a new subclass of `GDClientException` in the `com.github.alex1304.jdash.exception` package (can be the same as the one you create for demon rating)

File: src/main/java/com/github/alex1304/jdash/client/GDUserBlockRequest.java
Comment (Alex1304) [inline src/main/java/com/github/alex1304/jdash/client/GDUserBlockRequest.java:30]: Same here

File: src/main/java/com/github/alex1304/jdash/client/GDUserUnblockRequest.java
Comment (Alex1304) [inline src/main/java/com/github/alex1304/jdash/client/GDUserUnblockRequest.java:30]: And here

File: src/main/java/com/github/alex1304/jdash/exception/BadResponseException.java
Comment (Alex1304) [inline src/main/java/com/github/alex1304/jdash/exception/BadResponseException.java:16]: why is this change needed ? The default `toString()` of `Throwable` is just fine imo

File: src/main/java/com/github/alex1304/jdash/util/Utils.java
Comment (Alex1304) [inline src/main/java/com/github/alex1304/jdash/util/Utils.java:153]: This should never happen so you can `throw new AssertionError()`. Not a good practice to swallow errors like that.

**Review by Alex1304 — CHANGES_REQUESTED** [review]
Overall good, besides some adjustments.

File: src/main/java/com/github/alex1304/jdash/util/Utils.java
Comment (acceler8tion) [reply src/main/java/com/github/alex1304/jdash/util/Utils.java]: Changed to throwing AssertionError 60055b5

File: src/main/java/com/github/alex1304/jdash/exception/BadResponseException.java
Comment (acceler8tion) [reply src/main/java/com/github/alex1304/jdash/exception/BadResponseException.java]: Reverted 5888910

File: src/main/java/com/github/alex1304/jdash/client/AbstractGDClient.java
Comment (acceler8tion) [reply src/main/java/com/github/alex1304/jdash/client/AbstractGDClient.java]: Reverted 5888910

File: src/test/java/com/github/alex1304/jdash/TestMain.java
Comment (acceler8tion) [reply src/test/java/com/github/alex1304/jdash/TestMain.java:131]: I changed it because it was a little awkward, but should I change it to the original?

File: src/main/java/com/github/alex1304/jdash/client/AuthenticatedGDClient.java
Comment (acceler8tion) [reply src/main/java/com/github/alex1304/jdash/client/AuthenticatedGDClient.java]: Renamed to getFriends 14270ba

The reason why method name was `getPrivateFriends` was named after `getPrivateMessages`, but it was a bit strange to think about it again.

**Comment by acceler8tion** [comment]
@Alex1304 
I've created 4 sub-classes for "GDClientException," 
but I don't know what they need and how to add them.

File: src/main/java/com/github/alex1304/jdash/exception/BadResponseException.java
Comment (Alex1304) [reply src/main/java/com/github/alex1304/jdash/exception/BadResponseException.java]: you forgot to remove the toString() override you added

File: src/test/java/com/github/alex1304/jdash/TestMain.java
Comment (Alex1304) [reply src/test/java/com/github/alex1304/jdash/TestMain.java:131]: as i said it doesn't have any effect so it doesn't matter, just letting you know that the order of the operators don't change anything

File: src/main/java/com/github/alex1304/jdash/exception/BadResponseException.java
Comment (acceler8tion) [reply src/main/java/com/github/alex1304/jdash/exception/BadResponseException.java]: oof my mistake

deleted 2f1776c

**Comment by Alex1304** [comment]
They don't need anything, besides extending GDClientException.

File: src/main/java/com/github/alex1304/jdash/client/GDLevelStarsRatingRequest.java
Comment (acceler8tion) [reply src/main/java/com/github/alex1304/jdash/client/GDLevelStarsRatingRequest.java:46]: Added f5d254d

File: src/main/java/com/github/alex1304/jdash/client/GDUserUnblockRequest.java
Comment (acceler8tion) [reply src/main/java/com/github/alex1304/jdash/client/GDUserUnblockRequest.java:30]: Added f5d254d

File: src/main/java/com/github/alex1304/jdash/client/GDUserBlockRequest.java
Comment (acceler8tion) [reply src/main/java/com/github/alex1304/jdash/client/GDUserBlockRequest.java:30]: Added f5d254d

File: src/main/java/com/github/alex1304/jdash/client/GDLevelDemonRatingRequest.java
Comment (Alex1304) [reply src/main/java/com/github/alex1304/jdash/client/GDLevelDemonRatingRequest.java:36]: well you're supposed to use them here in place of your IllegalArgumentExceptions

Comment (acceler8tion) [reply src/main/java/com/github/alex1304/jdash/client/GDLevelDemonRatingRequest.java]: i think i'm crazy.. b5d278e

**Review by Alex1304 — APPROVED** [review]

**Comment by Alex1304** [comment]
Thanks. i'll do some testing tomorrow before merging.

**Comment by acceler8tion** [comment]
ok alex :+1:

**Comment by Alex1304** [comment]
Did you have the opportunity to investigate more about demon rating returning 500 error ? I have the issue too, probably wrong or missing parameters ?

**Comment by acceler8tion** [comment]
According to packet capture, there's no problem. `500 Internal Error` happened earlier this year, and it still hasn't been fixed. i swear

In fact, one of the things I've suspected of being a RobTop(i mean gd) is about this function, and Damon Rating is back in a voting state where you can go out of `Saved` and come back in.(Of course, you can't re-vote like and star rating.)

This occurance was also present when 500 Internal Error occurred in the past.

**Comment by Alex1304** [comment]
you mean the 500 error happens in-game too? Well in that case I guess it's "normal" then?

**Comment by acceler8tion** [comment]
in game too

**Comment by Alex1304** [comment]
I'm just gonna merge this and fix at a later date if more issues are found (don't hesitate to create issues as well if that happens). For the 500 error I'll consider it normal for now.

**Pull request merged**
